src/Week4/HW4.hs

```diff
@@ -0,0 +1,133 @@
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE LambdaCase #-}
+module Week4.HW4 where
+
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+-- Define the representation of the polynomial x (coeff = 1, degree = 1)
+x :: Num a => Poly a
+x = P [0,1]
+
+-- Exercise 2 ----------------------------------------
+
+-- Implement the (==) function
+
+-- This is a bit like list comparison, except that P[a,b,0,...,0] == P[a,b]
+-- i.e., we need to ignore trailing zeroes
+-- this can be done through pattern matching:
+--instance (Num a, Eq a) => Eq (Poly a) where
+--    (==) (P (a:as)) (P (b:bs)) = a == b && (P as) == (P bs)
+--    (==) (P []) (P (b':bs')) = b' == 0 && (P []) == (P bs')
+--    (==) (P(a':as')) (P []) = a' == 0 && (P as') == (P [])
+--    (==) (P [])(P[]) = True
+
+-- or, more concisely:
+instance (Num a, Eq a) => Eq (Poly a) where
+  (==) (P a) (P b) = (dropZeros $ reverse $ a) == (dropZeros $ reverse $ b)
+
+-- helpers
+-- Drops leading zeros from numeric list
+dropZeros :: (Num a, Eq a) => [a] -> [a]
+dropZeros = \case
+  a:as -> if a/=0 then (a:as) else dropZeros as
+  []   -> []
+
+-- function to reverse list (turns out we can just use `reverse` from Prelude)
+-- revList :: [a] -> [a]
+-- revList = foldl (\acc x -> x:acc) []
+
+-- Exercise 3 -----------------------------------------
+
+-- implement the show function for polynomials
+
+instance (Num a, Eq a, Show a) => Show (Poly a) where
+    show (P a) = if a == [] || all (==0) a then "0"
+                  else showTuple $ reverse $ toTuple a
+
+-- transforms list of a into tuple of (coeff, ord), and also gets rid of 0s
+toTuple :: (Num a, Eq a) => [a] -> [(a,a)]
+toTuple lst = helper lst 0
+  where
+   helper [] _       = []
+   helper (a:as) ord = if a == 0 then helper as (ord+1) else [(a,ord)] ++ helper as (ord+1)
+
+--shows (coeff, ord) list in a more readable way 
+--I tried to make the syntax prettier but without much sucess :(
+showTuple :: (Num a, Eq a, Show a) => [(a,a)] -> String
+showTuple []     = ""
+showTuple ((a,ord):as) = case ord of
+  0 -> show a ++ rest
+  1 -> if a == 1          then "x" ++ rest 
+        else if a == (-1) then "-x" ++ rest
+         else show a ++ "x" ++ rest
+  _ -> if a == 1          then "x^" ++ show ord ++ rest
+        else if a == (-1) then "-x^" ++ show ord ++ rest 
+         else show a ++ "x^" ++ show ord ++ rest
+  where
+   rest = if length as > 0 then " + " ++ showTuple as else ""
```

Comment:
length takes linear time, and you're checking length recursively for each tail of `as`, which leads to quadratic  algorithm in the end. You  could avoid that by using something like `not (null as)` which takes constant time, but probably better approach is to take a step back and use [intercalate](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List.html#v:intercalate) to insert " + " between all nonempty terms, so you could avoid explicit recursion.

----------------
Also using guard syntax is more idiomatic haskell, so instead of 

```haskell 
case x of
    1 -> if cond1 then ... 
         else if cond2 then ...
         else ...
    2 ->
```
you can write
```haskell
case x of
   1 | cond1 -> ...
     | cond2 -> ...
     | otherwise -> ...
   2 ->
```
which is arguably more readable.

Reply:
This makes sense! Thanks